third_party/move/move-model/src/builder/exp_builder.rs

```diff
@@ -2328,6 +2383,33 @@ impl<'env, 'translator, 'module_translator> ExpTranslator<'env, 'translator, 'mo
         }
     }
 
+    fn translate_typed_lvalue(
+        &mut self,
+        tlv: &EA::TypedLValue,
+        expected_type: &Type,
+        expected_order: WideningOrder,
+        match_locals: bool,
+        context: &ErrorMessageContext,
+    ) -> Pattern {
+        use move_ir_types::sp;
+        let sp!(loc, EA::TypedLValue_(lv, opt_ty)) = tlv;
+        match opt_ty {
+            Some(ty) => {
+                let loc = self.to_loc(loc);
+                let bound_type = self.translate_type(ty);
+                self.check_type_with_order(
+                    expected_order,
+                    &loc,
+                    &bound_type, // is this arg ordering right?
+                    expected_type,
+                    context,
+                );
+                self.translate_lvalue(lv, &bound_type, expected_order, match_locals, context)
```

Comment:
Why are we using the `expected_order` here? Shouldn't we check tha lvalues have the `bound_type` exactly, similar to how wo check annotated expressions?
```
EA::Exp_::Annotate(exp, typ) => {
	let ty = self.translate_type(typ);
	let exp =
		self.translate_exp_in_context(exp, &ty, &ErrorMessageContext::TypeAnnotation);
	self.check_type(&loc, &ty, expected_type, context);
	exp
},
```

Reply:
I don't think so.  The type of the `LValue` is the type of a container, not a value.  It may be wider than the value put in it.  I've added a test `lambda_typed_widen.move` which tries to use the "subtyping" relationship between `&mut u64` and `& u64` to test what happens to lambda params with declared types versus those with undeclared types.  Not sure it's a strong enough test, though.